Speed up `Refaster` bug checker

Stephan202 opened the pull request:
:exclamation: ~~This PR is on top of #255~~ :exclamation:

This PR is a work in progress; @rickie and I have worked on this on-and-off over the past months. Pulling this change across the finish line is a requirement for enabling the `Refaster` bug checker by default in our internal code base.

Open points include general code review, resolving various comments, and increasing test coverage.

This PR will also require changes in the context of #255, as that PR introduces a new type of `CodeTransformer`. Since that PR also introduces `RefasterTest`, it might be better to finalize that PR first, and apply this PR on top of the result.

First proposal of a suggested commit message:
```
Speed up `Refaster` bug checker (#261)

By not applying Refaster templates to the compilation unit under consideration,
if a failure to match can efficiently be established ahead of time.

The approach taken comprises a two-step process:
1. Upon loading Refaster templates from the classpath, inspect their structure
   and extract, for each possible variant (as defined by a `@BeforeTemplate`
   method or `Refaster#anyOf` usage), a set of syntactical components that must
   be present for the template to match. This includes compile-time constant
   strings, as well as various operators.
2. When analyzing a compilation unit, perform the same type of "syntactical
   component extraction" and then attempt to apply only those Refaster
   templates that contain a subset of the extracted components.

The first step indexes Refaster templates into an efficient tree data
structure, such that the second step can efficiently look up candidate Refaster
templates.

This change provides a nice speed-up in the common case, and reduces the
performance impact of any newly introduced Refaster template. For example, the
large battery of Refaster templates matching against one of AssertJ's
`Assertions#assertThat` overloads are now no longer applied to non-test code.
```

Stephan202 committed:
WIP: Speed up `Refaster` check

This code assumes a modification of Error Prone; see the
`sschroevers/refaster-tree-tweaks` branch on the fork.

Stephan202 committed:
Compatibility with stock Error Prone

Stephan202 committed:
Emit website link along with Refaster refactor suggestions

Stephan202 committed:
More extensible approach

Stephan202 committed:
WIP: Some plumbing for annotation support

To be used for custom links, custom error messages, custom other stuff...

Stephan202 committed:
Another round

rickie committed:
Initial copy over of the improved algorithm

rickie committed:
Improve code and algorithm for Refaster

Stephan202 committed:
Tweaks

rickie committed:
Drop unnecessary `@AutoService` annotation

rickie committed:
Add XXXs

Stephan202 committed:
Better annotation support, simplify setup

Stephan202 committed:
Flag why build currently doesn't fail, while it should.

Stephan202 committed:
Improve logic and test coverage

Stephan202 committed:
Properly document URL placeholder usage

Stephan202 committed:
Expand test coverage

Stephan202 committed:
Flag classpath issue

Stephan202 committed:
Merge `RefasterRuleSelector` type hierarchy

Stephan202 committed:
Create selector only once per `Refaster` instantiation

Stephan202 committed:
Move all `RefasterRuleSelector` construction logic into the relevant class

Stephan202 committed:
Push sorting requirements into `Node`, minimize tree, add tests

rickie reviewed (commented):
Have one tiny suggestion for renaming. Will stop reviewing to prioritize the Refaster annotation PR first.

rickie committed:
Suggestion

rickie committed:
Extract the `TreeScanner`s to their own classes

rickie committed:
Introduce `getClass` method to deduplicate

rickie committed:
Reorder methods in `RefasterIntrospection`

rickie committed:
Improve text and minor improvements

rickie committed:
Use `@AutoValue` for the `AnnotatedCompositeCodeTransformer`

rickie committed:
Support `AllSuggestionsAsWarnings` and add a suggestion

rickie committed:
Tweak

Stephan202 committed:
Suggestions, introduce `ErrorProneFork`

Stephan202 committed:
Also this

Stephan202 commented on refaster-runner/src/test/java/tech/picnic/errorprone/refaster/runner/NodeTest.java:
```diff
@@ -0,0 +1,177 @@
+  @MethodSource("verifyTestCases")
+  @ParameterizedTest
+  void verify(ImmutableSetMultimap<Integer, ImmutableSet<String>> source, Random random) {
+    Node<Integer> tree = Node.create(source.keySet().asList(), source::get);
```
Maybe `treeInputSource` indeed. Will sleep on it :smile:

Stephan202 commented on refaster-runner/src/main/java/tech/picnic/errorprone/refaster/runner/RefasterRuleSelector.java:
```diff
@@ -0,0 +1,516 @@
+   * @return A string representation of the operator, if known
+   * @throws IllegalArgumentException If the given input is not supported.
+   */
+  // XXX: Extend list to cover remaining cases; at least for any `Kind` that may appear in a
```
Additionally, we should eventually perhaps do something similar for certain keywords (`instanceof`, `new`, ...), and possibly some other things. Might be best to do that once we have some benchmarking in place.

Stephan202 reviewed (commented):
Rebased.

Stephan202 committed:
Clarify how the default Refaster hit severity comes to be

Plus other tweaks.

Badbond committed:
Align documentation of reported description names by the Refaster check

Badbond committed:
Suggestion

Badbond committed:
Pass null for urlLink to Description.Builder

ibabiankou reviewed (commented):
Initial suggestions. Only looked at `Node` and `Refaster` will continue next week. Also pushed a commit, PTAL.

Stephan202 committed:
Tweaks

Stephan202 committed:
Apply some suggestions

Stephan202 reviewed (commented):
Thanks for the review @ibabiankou!

> Also pushed a commit, PTAL.

Didn't see it :eyes: (But I did just push something :smile: .)

Stephan202 committed:
Optimize code, introduce benchmark, simplify test

Stephan202 committed:
Comment style, explain both performance-only pieces of code

Since PIT correctly flags these as "redundant".

Also for JDK 11 compatibility.

rickie committed:
Move `AnnotatedCompositeCodeTransformer` and `ErrorProneFork` to `refaster-support`

rickie committed:
`s/information/content/`

rickie committed:
Improve Javadoc `AnnotatedCompositeCodeTransformer`

rickie committed:
Revert changes in `OptionalTemplates`

Badbond committed:
Tweak `AnnotatedCompositeCodeTransformer` Javadoc

rickie committed:
Fix typo

ibabiankou commented on refaster-runner/src/main/java/tech/picnic/errorprone/refaster/runner/Node.java:
```diff
@@ -0,0 +1,111 @@
+        child.collectReachableValues(remainingCandidateEdges, sink);
+      }
+      collectReachableValues(remainingCandidateEdges, sink);
+    }
```
> I see that in the internal code candidateEdges is the result of ImmutableSortedSet#asList(), such that the contains check is O(1), while here it's O(n). 

Ah, in that case, it makes more sense 😄 I will have a closer look at the benchmark. 👍

ibabiankou commented on refaster-runner/src/main/java/tech/picnic/errorprone/refaster/runner/RefasterRuleSelector.java:
```diff
@@ -0,0 +1,554 @@
+ *       </ol>
+ *   <li>Extract all identifiers from the {@link CompilationUnitTree} and sort them
+ *       lexicographically.
+ *   <li>Traverse the tree based on the identifiers from the {@link CompilationUnitTree}. Every node
+ *       can contain Refaster templates. Once a node is we found a candidate Refaster template and
+ *       will therefore be added to the list of candidates.
```
```suggestion
 *   <li>Traverse the tree based on the identifiers from the {@link CompilationUnitTree}. Every node
 *       can contain Refaster templates. Once a node is we found a candidate Refaster template that
 *       might match some code and will therefore be added to the list of candidates.
```

ibabiankou commented on refaster-runner/src/main/java/tech/picnic/errorprone/refaster/runner/RefasterRuleSelector.java:
```diff
@@ -0,0 +1,554 @@
+ *        └── D     -- T3
+ * }</pre>
+ *
+ * <p>The tree is traversed based on the identifiers in the {@link CompilationUnitTree}. When a leaf
+ * contains a template and is reached, we can be certain that the identifiers from the {@link
```
```suggestion
 * <p>The tree is traversed based on the identifiers in the {@link CompilationUnitTree}. When a node
 * containing a template is reached, we can be certain that the identifiers from the {@link
```

ibabiankou commented on refaster-runner/src/main/java/tech/picnic/errorprone/refaster/runner/RefasterRuleSelector.java:
```diff
@@ -0,0 +1,554 @@
+ * contains a template and is reached, we can be certain that the identifiers from the {@link
+ * BeforeTemplate} are at least present in the {@link CompilationUnitTree}.
+ *
+ * <p>Since the identifiers are sorted, we can prune parts of the {@link Node tree} while we are
```
Prune implies modification of the tree.
```suggestion
 * <p>Since the identifiers are sorted, we can skip parts of the {@link Node tree} while we are
```

ibabiankou commented on refaster-runner/src/main/java/tech/picnic/errorprone/refaster/runner/RefasterRuleSelector.java:
```diff
@@ -0,0 +1,554 @@
+ *
+ * <p>Since the identifiers are sorted, we can prune parts of the {@link Node tree} while we are
+ * traversing it. Instead of trying to match all Refaster templates against every expression in a
+ * {@link CompilationUnitTree} we now only return a subset of the templates that at least have a
+ * chance of matching. As a result, the performance of Refaster significantly increases.
```
```suggestion
 * {@link CompilationUnitTree} we now only matching a subset of the templates that at least have a
 * chance of matching. As a result, the performance of Refaster increases significantly.
```

ibabiankou commented on refaster-runner/src/main/java/tech/picnic/errorprone/refaster/runner/RefasterRuleSelector.java:
```diff
@@ -0,0 +1,554 @@
+        getMethod(UCLASS_IDENT, "getTopLevelClass");
+    private static final Method METHOD_UANY_OF_EXPRESSIONS = getMethod(UAnyOf.class, "expressions");
+
+    static boolean isUClassIdent(IdentifierTree tree) {
```
Let's use the full word.
```suggestion
    static boolean isUClassIdentifier(IdentifierTree tree) {
```

ibabiankou commented on refaster-runner/src/main/java/tech/picnic/errorprone/refaster/runner/RefasterRuleSelector.java:
```diff
@@ -0,0 +1,554 @@
+    }
+
+    // Actually UClassIdent.
+    static IdentifierTree getClassIdent(UStaticIdent tree) {
```
```suggestion
    static IdentifierTree getClassIdentifier(UStaticIdent tree) {
```

ibabiankou commented on refaster-runner/src/main/java/tech/picnic/errorprone/refaster/runner/RefasterRuleSelector.java:
```diff
@@ -0,0 +1,554 @@
+
+    private static void registerOperator(
+        ExpressionTree node, List<Set<String>> identifierCombinations) {
+      identifierCombinations.forEach(ids -> ids.add(treeKindToString(node.getKind())));
```
The identifier does not change.
```suggestion
      String id = treeKindToString(node.getKind());
      identifierCombinations.forEach(ids -> ids.add(id));
```

ibabiankou reviewed (commented):
A few more comments and suggestions.

Stephan202 committed:
Apply `StringJoin` suggestion

ibabiankou committed:
Suggestions

Stephan202 committed:
Suggestions

Stephan202 committed:
Suggestions

Stephan202 commented:
Rebased and added a small commit; recent changes LGTM.

rickie committed:
Fix typo

Stephan202 committed:
Merge remote-tracking branch 'origin/sschroevers/refaster-custom-urls' into sschroevers/refaster-speed-up

Stephan202 committed:
Introduce `AnnotatedCompositeCodeTransformerTest`

Stephan202 commented on refaster-runner/src/main/java/tech/picnic/errorprone/refaster/runner/Refaster.java:
```diff
@@ -146,16 +159,40 @@ private static Stream<Replacement> getReplacements(
+      throw new IllegalStateException(
+          String.format("Can't handle `CodeTransformer` of type '%s'", transformer.getClass()));
+    }
   }
```
@ibabiankou I agree that with the PR as-is it was an improvement to move this `RefasterRule` extraction logic here, but the changes in #255 make the `CodeTransformer` -> `ImmutableSet<ImmutableSet<String>>` transformation less straightforward, so I'll move it back. Of course, idea on how to better do this are welcome!

Stephan202 committed:
Merge remote-tracking branch 'origin/sschroevers/refaster-custom-urls' into sschroevers/refaster-speed-up

Stephan202 reviewed (commented):
Merged the changes of #255 into this PR. Will change the base branch.

Stephan202 committed:
Suggestions

Stephan202 reviewed (commented):
Added a commit. More review TBD.

rickie reviewed (commented):
One small comment. Code improvements are cool 🚀 !

Stephan202 commented on refaster-runner/src/main/java/tech/picnic/errorprone/refaster/runner/RefasterRuleSelector.java:
```diff
@@ -320,12 +315,11 @@ static ImmutableList<UStatement> getTemplateStatements(BlockTemplate template) {
     }
 
-    // XXX: Make nicer. Or rename the other params.
+    // Arguments to this method must actually be of the package-private type `UClassIdent`.
```
One passes arguments _to_ a method. (Corresponding to the parameters _of_ said method. But here we are talking about the provided argument; the comment is here precisely because the parameter type isn't specific enough.)

Stephan202 committed:
Merge branch 'master' into sschroevers/refaster-speed-up

Stephan202 commented:
Resolved a whole bunch of conflicts. I did need to suppress some `ErrorProneRuntimeClasspath` false positives; that should be tackled in a separate PR.

rickie committed:
Merge branch master into sschroevers/refaster-speed-up

rickie committed:
Merge branch 'master' into sschroevers/refaster-speed-up

rickie committed:
Post-merge-fix

rickie commented:
/integration-test

rickie commented:
I updated the branch with master, now running the integration tests, so lets see how that goes!

rickie committed:
Merge `master` into `sschroevers/refaster-speed-up`

rickie committed:
Merge branch 'master' into sschroevers/refaster-speed-up

rickie committed:
Merge branch master into sschroevers/refaster-speed-up

rickie committed:
Add `Refaster#AnyOf` use cases

rickie committed:
Run other IT although Checkstyle fails

rickie commented:
/integration-test

rickie committed:
Continue running if one IT fails

rickie commented:
/integration-test

rickie committed:
Another try to keep it running

rickie commented:
/integration-test

rickie committed:
Try to drop it then

rickie committed:
Revert this after testing

rickie committed:
Merge branch 'master' into sschroevers/refaster-speed-up

rickie committed:
Merge `master` into this branch

rickie committed:
Apply suggestions

rickie committed:
Merge `master` in the speed up branch

rickie committed:
Merge branch 'master' into sschroevers/refaster-speed-up

rickie committed:
Merge `master` and disable a rule temporarily

rickie committed:
Revert later: temporary introduce jitpack yml

rickie committed:
Merge `master` in branch

rickie committed:
Merge branch 'master' into sschroevers/refaster-speed-up

rickie committed:
Post-merge fixes and improvements

rickie committed:
Revert some old changes now that the pom is correct again

rickie committed:
Merge `master` into branch

rickie committed:
Tweaks

rickie committed:
Kill mutants in `Node`

rickie committed:
Some were not covering the mutant...

rickie reviewed (commented):
I think getting rid of all mutations will be hard 🤔.

The state of the PR is already looking quite good!

rickie committed:
Add more testing for visitor

rickie committed:
Add a test case back

rickie committed:
Add blocktemplate test

rickie committed:
Add some more to kill more

rickie committed:
Revert "Add some more to kill more"

This reverts commit b6170583b84192ca95bd02ea635afce4a23784e0.

rickie committed:
Revert "Add blocktemplate test"

This reverts commit c5237cc2e259db752e79bdd80b36efc5c7e5ef76.

rickie committed:
Drop another test as it doesnt have an impact

rickie committed:
Introduce `RefasterStrategy` and tests

rickie committed:
Extract `SourceIdentifierExtractor` and test it

rickie committed:
Initial impl of the extractor

rickie committed:
Cleanup and more coverage

rickie committed:
Drop comments

rickie committed:
Update examples to satisfy selfcheck

rickie committed:
Add new test

rickie committed:
Fix test

rickie committed:
Further cleanup

rickie committed:
Further improvements

rickie committed:
More cleanup and more tests

rickie committed:
Add more tests

rickie committed:
One step further and drop variable names

rickie committed:
Two more cases to show they work

rickie committed:
Use parameterized test

rickie committed:
Sort annotation

rickie committed:
Merge branch 'master' into sschroevers/refaster-speed-up

Stephan202 committed:
Merge remote-tracking branch 'origin/master' into sschroevers/refaster-speed-up

Stephan202 committed:
Suggestions

Stephan202 committed:
Merge branch 'master' into sschroevers/refaster-speed-up

Stephan202 committed:
Drop dependency declaration

Stephan202 commented:
Pushed a change with some tweaks and comments, already discussed offline with @rickie. Also resolved a conflict.

Stephan202 committed:
Suggestions

rickie committed:
Merge branch 'master' into sschroevers/refaster-speed-up

mohamedsamehsalah commented on refaster-runner/src/main/java/tech/picnic/errorprone/refaster/runner/RefasterRuleIdentifierExtractor.java:
```diff
@@ -0,0 +1,264 @@
+    return identifierCombinations.stream().map(ImmutableSet::copyOf).collect(toImmutableSet());
+  }
+
+  private static final class RefasterIntrospection {
```
Updating this will be annoying if `error-prone` change any of the method namings, but no strong reason to expect this.

mohamedsamehsalah reviewed (commented):
Smart work 🧠 🧠 🧠  — I don’t have much to add on the algorithm itself.

What I would love to see, as a _humble_ student of what is being done here, is to add extensive documentation on the classes and the methods; but I am sure this is part of the long list of XXX's 😄

rickie committed:
Merge branch 'master' into sschroevers/refaster-speed-up

rickie commented:
> What I would love to see, as a humble student of what is being done here, is to add extensive documentation on the classes and the methods; but I am sure this is part of the long list of XXX's 😄

I started working on that @mohamedsamehsalah !

rickie committed:
Fix illegal symbol Checkstyle error

rickie committed:
Document the `refaster-runner` module

rickie committed:
Rename `TreeKindUtil` to `TreeKindStringifier`

rickie committed:
Elaborate on optimized rule selection in the `refaster-runner` README

rickie commented:
One nit on the `forbiddenapis` change in `pom.xml`, @Stephan202: disabling `jdk-reflection` globally widens the blast radius of the relaxation across every module, even though in practice the only `setAccessible` user introduced here is `RefasterRuleIdentifierExtractor$RefasterIntrospection`.

Suggested tweak: re-enable `jdk-reflection` in the global bundle list, and exclude just the one class via the plugin's `<excludes>`:

```xml
<bundledSignatures>
    <bundledSignature>jdk-internal</bundledSignature>
    <bundledSignature>jdk-reflection</bundledSignature>
    <bundledSignature>jdk-system-out</bundledSignature>
</bundledSignatures>
<excludes>
    <exclude>tech/picnic/errorprone/refaster/runner/RefasterRuleIdentifierExtractor$RefasterIntrospection.class</exclude>
</excludes>
```

That way any future accidental `setAccessible` in the rest of the codebase still trips the check, while the intentional reflection into package-private Refaster types stays allowed. Verified locally: forbiddenapis reports 0 errors for the `refaster-runner` module with `jdk-reflection` active.

_Written with Claude_ 🤖

rickie committed:
Suggestions

Summary of changes:
- Re-enable the `jdk-reflection` forbiddenapis bundle and narrow the
  allowance to `RefasterIntrospection` via `<excludes>`, instead of
  disabling the bundle repo-wide.
- Add `@ParameterizedTest` to `RefasterTest#replacement` so the
  optimized-vs-disabled-strategy parity test actually runs.
- Delete the unused `RefasterStrategy` interface.
- Rename `NodeTest#collectReachableValuesWithEmptyChildren` to
  `collectReachableValuesWithValueAtRoot`, matching what it verifies.
- Convert `OptimizedCodeTransformer` to a record and document why its
  `annotations()` returns an empty map.
- Fix grammar and align the `T`/`R` template identifiers in
  `RefasterRuleSelector`'s class Javadoc with the README.
- Cross-reference the stay-in-sync invariant between
  `SourceIdentifierExtractor` and `RefasterRuleIdentifierExtractor`.

rickie committed:
Drop `NullAway` suppression on `RefasterRuleSelector#create`

`Map#get` is `@Nullable`, while `Node#create` expects a non-null
`Function` return. Replace the method reference with a lambda that
wraps the lookup in `requireNonNull`, which is safe because the keys
come from `Map#keySet`.

rickie committed:
Merge remote-tracking branch 'origin/master' into sschroevers/refaster-speed-up

# Conflicts:
#	pom.xml

mohamedsamehsalah commented on refaster-runner/README.md:
```diff
@@ -0,0 +1,47 @@
+# Refaster runner
```
❤️

mohamedsamehsalah commented on refaster-runner/README.md:
```diff
@@ -0,0 +1,47 @@
+
+## What does this module do?
+
+The `Refaster` bug checker scans the classpath for `.refaster` files,
```
I had to research this, but maybe elaborate on what exactly are `.refaster` files? (Maybe a reference to https://errorprone.info/docs/refaster)

mohamedsamehsalah commented on refaster-runner/README.md:
```diff
@@ -20,15 +20,63 @@ Two compiler flags tune this behavior:
+     extractor emits one identifier set per branch, so a single rule
+     may occupy multiple paths in the tree. Reaching any one of them
+     during traversal suffices to select the rule.
+   - Each set is sorted lexicographically and inserted into an
```
What's the functional reason for lexicographical sorting?

mohamedsamehsalah commented on refaster-runner/README.md:
```diff
@@ -20,15 +20,63 @@ Two compiler flags tune this behavior:
+     `TreeKindStringifier`.
+   - `Refaster.anyOf(a, b, c)` is treated as a disjunction: the
+     extractor emits one identifier set per branch, so a single rule
+     may occupy multiple paths in the tree. Reaching any one of them
```
Shall we introduce the concept of the "tree" before this? 😄

mohamedsamehsalah commented on refaster-runner/README.md:
```diff
@@ -39,9 +87,17 @@ Given three rules with identifier sets `R1 = [A, B, C]`, `R2 = [B]`, and `R3 =
+candidate rules are then matched and fixed exactly as in the
+unoptimized path, and overlapping replacements are resolved by
+`Refaster#applyMatches` as usual. See `RefasterRuleSelector` for the
+entry point.
```
> See `RefasterRuleSelector` for the entry point.

I feel like this should be in the "How it works" section.

mohamedsamehsalah reviewed (commented):
Thanks for the `readme` @rickie . I have some questions/thoughts.

rickie committed:
Suggestions
